### PR TITLE
[stable/kong] Add a chart for Kong

### PR DESCRIPTION
**Pull request by shashiranjan84**
- Chart deploys Kong and its backing datastore either Cassandra or
  PostgreSql(default)
- Uses official [Kong](https://hub.docker.com/_/kong/) Docker image

Note: Opening this new PR instead of #3031.

### CODE REVIEW

**Comment by shashiranjan84** [comment]
I have updated the the labels and made few more changes as advised in best practices guide.

File: stable/kong/Chart.yaml [hunk 1]
```diff
@@ -0,0 +1,13 @@
...
+engine: gotpl
+home: https://getkong.org/
+icon: https://s3.amazonaws.com/downloads.kong/universe/assets/icon-kong-inc-large.png
+maintainers:
+- name: shashiranjan84
+  email: shashi@konghq.com
+name: kong
+sources:
+- https://github.com/Kong/kong
+version: 0.1.0
```
Comment (unguiculus) [inline stable/kong/Chart.yaml:13]: Add `appVersion` field.

File: stable/kong/templates/NOTES.txt [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+   {{- else if contains "NodePort" .Values.admin.type }}
+     HOST=$(kubectl get nodes --namespace {{ .Release.Namespace }} -o jsonpath='{.items[0].status.addresses[0].address}')
+     PORT=$(kubectl get svc --namespace {{ .Release.Namespace }} {{ template "kong.fullname" . }}-admin -o jsonpath='{.spec.ports[0].nodePort}')
+
+   {{- else if contains "ClusterIP" .Values.admin.type }}
+     HOST=127.0.0.1
+
+     {{- if .Values.admin.https }}
+     # Execute the following commands to route the connection to Admin SSL port:
+     export POD_NAME=$(kubectl get pods --namespace {{ .Release.Namespace }} -l "app={{ template "kong.name" . }}" -o jsonpath="{.items[0].metadata.name}")
```
Comment (unguiculus) [inline stable/kong/templates/NOTES.txt]: Add `release` label to selector.

File: stable/kong/templates/NOTES.txt [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+     kubectl port-forward $POD_NAME {{ .Values.admin.https.servicePort }}:{{ .Values.admin.https.servicePort }}
+     {{- else if .Values.admin.http }}
+     # Execute the following commands to route the connection to Admin port:
+     export POD_NAME=$(kubectl get pods --namespace {{ .Release.Namespace }} -l "app={{ template "kong.name" . }}" -o jsonpath="{.items[0].metadata.name}")
```
Comment (unguiculus) [inline stable/kong/templates/NOTES.txt]: Add `release` label to selector.

File: stable/kong/templates/NOTES.txt [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+   {{- else if contains "NodePort" .Values.proxy.type }}
+     HOST=$(kubectl get nodes --namespace {{ .Release.Namespace }} -o jsonpath='{.items[0].status.addresses[0].address}')
+     PORT=$(kubectl get svc --namespace {{ .Release.Namespace }} {{ template "kong.fullname" . }}-proxy -o jsonpath='{.spec.ports[0].nodePort}')
+
+   {{- else if contains "ClusterIP" .Values.proxy.type }}
+     HOST=127.0.0.1
+
+     {{- if .Values.proxy.https }}
+     # Execute the following commands to route the connection to proxy SSL port:
+     export POD_NAME=$(kubectl get pods --namespace {{ .Release.Namespace }} -l "app={{ template "kong.name" . }}" -o jsonpath="{.items[0].metadata.name}")
```
Comment (unguiculus) [inline stable/kong/templates/NOTES.txt]: Add `release` label to selector.

File: stable/kong/templates/NOTES.txt [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+     kubectl port-forward $POD_NAME {{ .Values.proxy.https.servicePort }}:{{ .Values.proxy.https.servicePort }}
+     {{- else if .Values.proxy.http }}
+     # Execute the following commands to route the connection to proxy port:
+     export POD_NAME=$(kubectl get pods --namespace {{ .Release.Namespace }} -l "app={{ template "kong.name" . }}" -o jsonpath="{.items[0].metadata.name}")
```
Comment (unguiculus) [inline stable/kong/templates/NOTES.txt]: Add `release` label to selector.

File: stable/kong/templates/deployment.yaml [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+    component: "kong_server"
+spec:
+  replicas: {{ .Values.replicaCount }}
+  template:
+    metadata:
+    {{- if .Values.podAnnotations }}
+      annotations:
+{{ toYaml .Values.podAnnotations | indent 8 }}
+    {{- end }}
+      labels:
```
Comment (unguiculus) [inline stable/kong/templates/deployment.yaml:22]: Only `app` and ` release` labels should go here.

File: stable/kong/templates/migrations.yaml [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+    app: {{ template "kong.name" . }}
+    chart: "{{ .Chart.Name }}-{{ .Chart.Version }}"
+    release: "{{ .Release.Name }}"
+    heritage: "{{ .Release.Service }}"
+    component: "migrations"
+spec:
+  template:
+    metadata:
+      name: {{ template "kong.name" . }}-migrations
+      labels:
```
Comment (unguiculus) [inline stable/kong/templates/migrations.yaml:15]: Only `app` and ` release` labels should go here.

File: stable/kong/templates/deployment.yaml [hunk 8]
```diff
@@ -0,0 +1,86 @@
+apiVersion: extensions/v1beta1
+kind: Deployment
+metadata:
+  name: "{{ template "kong.fullname" . }}"
+  labels:
+    app: "{{ template "kong.name" . }}"
+    chart: "{{ .Chart.Name }}-{{ .Chart.Version }}"
+    release: "{{ .Release.Name }}"
+    heritage: "{{ .Release.Service }}"
+    component: "kong_server"
```
Comment (unguiculus) [inline stable/kong/templates/deployment.yaml]: It doesn't look like you have real components in your chart. There is only a single deployment. Usually, components are used to identify multiple deployments in a chart and then are added to label selectors. This is not the case here. Do they server any special purpose?

File: stable/kong/README.md [hunk 9]
```diff
@@ -0,0 +1,123 @@
...
+| affinity                          | Node/pod affinities                                                    |                       |
+| nodeSelector                      | Node labels for pod assignment                                         | `{}`                  |
+| podAnnotations                    | Annotations to add to each pod                                         | `{}`                  |
+| resources                         | Pod resource requests & limits                                         | `{}`                  |
+| tolerations                       | List of node taints to tolerate                                        | `[]`                  |
+
+### Kong-specific parameters
+
+Kong has a choice of either Postgres or Cassandra as a backend datatstore.
+This chart allows you to choose either of them with the `kong.database.type`
```
Comment (unguiculus) [inline stable/kong/README.md]: `kong.database.type` is not set anywhere but mentioned here.

File: stable/kong/values.yaml [hunk 10]
```diff
@@ -0,0 +1,268 @@
...
+# This chart allows you to choose either of them with the `database.type`
+# parameter.  Postgres is chosen by default.
+
+# Additionally, this chart allows you to use your own database or spin up a new
+# instance by using the `postgres.enabled` or `cassandra.enabled` parameters.
+# Enabling both will create both databases in your cluster, but only one
+# will be used by Kong based on the `env.database` parameter.
+# Postgres is enabled by default.
+
+# Cassandra chart configs
```
Comment (unguiculus) [inline stable/kong/values.yaml:102]: It doesn't make sense to reproduce the whole Cassandra and Postgres values files here.

File: stable/kong/values.yaml [hunk 11]
```diff
@@ -0,0 +1,268 @@
...
+  nodePort: 32443
+
+# Set runMigrations to run Kong migrations
+runMigrations: true
+
+# Specify Kong configurations
+# Kong configurations guide https://getkong.org/docs/latest/configuration/
+env:
+  database: postgres
+  pg_password: kong
```
Comment (unguiculus) [inline stable/kong/values.yaml]: You should not added a second field for the DB password. There is already `postgresql.postgresPassword`. Also, by default, a random password is generated. Let's keep it that way.

**Review by unguiculus — CHANGES_REQUESTED** [review]
Thanks for the PR. I did an initial review.

**Comment by shashiranjan84** [comment]
@unguiculus thanks for review. I will make the changes you advised. Please expect some delay as I am travelling and have limited internet connectivity.

**Comment by shashiranjan84** [comment]
@unguiculus please review.

File: stable/kong/README.md [hunk 12]
```diff
@@ -0,0 +1,123 @@
...
+```
+
+## Introduction
+
+This chart bootstraps all the components needed to run Kong on a [Kubernetes](http://kubernetes.io)
+cluster using the [Helm](https://helm.sh) package manager.
+
+## Prerequisites
+
+- Kubernetes 1.6+ with Beta APIs enabled.
```
Comment (unguiculus) [inline stable/kong/README.md]: 1.8+

We offically support the current and previous minor versions.

File: stable/kong/templates/deployment.yaml [hunk 13]
```diff
@@ -0,0 +1,88 @@
+apiVersion: extensions/v1beta1
```
Comment (unguiculus) [inline stable/kong/templates/deployment.yaml]: Use `apps/v1beta2`

File: stable/kong/values.yaml [hunk 14]
```diff
@@ -0,0 +1,113 @@
...
+# Cassandra chart configs
+cassandra:
+  enabled: false
+
+# PostgreSQL chart configs
+postgresql:
+  enabled: true
+  postgresUser: kong
+  postgresPassword: kong
+  postgresDatabase: kong
```
Comment (unguiculus) [inline stable/kong/values.yaml]: Comment this out so the password is auto-generated by default.

**Review by unguiculus — CHANGES_REQUESTED** [review]
Please don't squash commits. They will be squashed upon merge anyways.

File: stable/kong/templates/deployment.yaml [hunk 15]
```diff
@@ -0,0 +1,91 @@
...
+  labels:
+    app: "{{ template "kong.name" . }}"
+    chart: "{{ .Chart.Name }}-{{ .Chart.Version }}"
+    release: "{{ .Release.Name }}"
+    heritage: "{{ .Release.Service }}"
+spec:
+  replicas: {{ .Values.replicaCount }}
+  selector:
+    matchLabels:
+      app: {{ template "kong.name" . }}
```
Comment (unguiculus) [inline stable/kong/templates/deployment.yaml:14]: Add `release` label to selector.

**Review by unguiculus — CHANGES_REQUESTED** [review]

File: stable/kong/templates/deployment.yaml (hunk 15 above)
Comment (shashiranjan84) [reply stable/kong/templates/deployment.yaml:14]: Done.

**Comment by shashiranjan84** [comment]
@unguiculus are we still waiting on something?

**Comment by shashiranjan84** [comment]
ping @unguiculus

File: stable/kong/templates/migrations.yaml [hunk 16]
```diff
@@ -0,0 +1,45 @@
...
+        app: "{{ template "kong.name" . }}"
+        release: "{{ .Release.Name }}"
+    spec:
+      containers:
+      - name: {{ template "kong.name" . }}-migrations
+        image: "{{ .Values.image.repository }}:{{ .Values.image.tag }}"
+        imagePullPolicy: {{ .Values.image.pullPolicy }}
+        env:
```
Comment (unguiculus) [inline stable/kong/templates/migrations.yaml:23]: For consistency, don't indent this list.

**Review by unguiculus — CHANGES_REQUESTED** [review]

File: stable/kong/templates/migrations.yaml (hunk 16 above)
Comment (shashiranjan84) [reply stable/kong/templates/migrations.yaml:23]: makes sense, @unguiculus I fixed it.

**Pull request merged**